File: typescript/src/companion/session.ts
// companion/session.ts
// The conscious loop: a concrete ICompanionSession that recalls from fused
// memory, persists each turn, and encodes it into the graph off the hot path.
// Ported from Circle.AI.Companion (CompanionSession) — the C# reference.
//
// This is the wire that makes the memory-brain a brain instead of a stub: on
// every turn it (1) recalls the most relevant memories + the user's own facts
// and injects them into the system prompt, (2) calls the generator, (3) persists
// the exchange to episodic memory, and (4) hands it to the background encoder so
// the knowledge graph fills for future associative recall.

import type {
  CompanionContext,
  CompanionTurn,
  ICompanionSession,
  InterfaceKind,
  ProactiveMessageHandler,
} from "./index.js";
import type { ChatMessage, IChatGenerator } from "../inference/index.js";
import type { EpisodicMemoryEntry, IEpisodicMemoryStore } from "../memory/index.js";
import type { IRecall } from "../memory/recall.js";
import type { CompanionMemoryEncoder } from "./memory_encoder.js";
import type { SelfBeliefStore } from "./belief.js";
import { randomUUID } from "node:crypto";

/** Construction-time configuration for a {@link CompanionSession}. */
export interface CompanionSessionOptions {
  readonly sessionId: string;
  readonly identityId: string;
  readonly interface: InterfaceKind;
  readonly displayName?: string;
  readonly preferredLanguage?: string | null;
  /** Static persona hint block prepended to the system prompt. */
  readonly personaHints?: string;
  /** Static affect hint block prepended to the system prompt. */
  readonly affectSummary?: string;
  readonly activeGoals?: readonly string[];
  /** How many memories to recall per turn. Default 5. */
  readonly recallTopK?: number;
  /** Optional app context stamped onto persisted episodes. */
  readonly appContext?: string;
  /** Background graph/belief encoder. When null, turns are not encoded. */
  readonly encoder?: CompanionMemoryEncoder | null;
  /** The user's own facts, surfaced into the system prompt. */
  readonly beliefs?: SelfBeliefStore | null;
  /** Optional embedder for associative episodic recall; null → recency recall. */
  readonly embedder?: ((text: string) => Promise<number[] | null>) | null;
}

/** A companion session that thinks with fused memory and remembers what it learns. */
export class CompanionSession implements ICompanionSession {
  readonly sessionId: string;
  readonly identityId: string;
  readonly interface: InterfaceKind;
  onProactiveMessageReady: ProactiveMessageHandler | null = null;

  private readonly generator: IChatGenerator;
  private readonly episodic: IEpisodicMemoryStore;
  private readonly recall: IRecall;
  private readonly opts: CompanionSessionOptions;
  private readonly _history: CompanionTurn[] = [];
  private _context: CompanionContext;

  constructor(
    generator: IChatGenerator,
    episodic: IEpisodicMemoryStore,
    recall: IRecall,
    opts: CompanionSessionOptions,
  ) {
    if (!generator) throw new Error("generator required");
    if (!episodic) throw new Error("episodic required");
    if (!recall) throw new Error("recall required");
    this.generator = generator;
    this.episodic = episodic;
    this.recall = recall;
    this.opts = opts;
    this.sessionId = opts.sessionId;
    this.identityId = opts.identityId;
    this.interface = opts.interface;
    this._context = this.buildContext([]);
  }

  get history(): readonly CompanionTurn[] {
    return this._history;
  }

  async sendAsync(message: string): Promise<string> {
    const prepared = await this.prepareAsync(message);
    const reply = await this.generator.generateAsync(prepared.messages);
    await this.recordTurnAsync(message, reply, prepared.queryEmbedding, prepared.snippets);
    return reply;
  }

  async *streamAsync(message: string): AsyncGenerator<string> {
    const prepared = await this.prepareAsync(message);
    let reply = "";
    for await (const chunk of this.generator.streamAsync(prepared.messages)) {
      reply += chunk;
      yield chunk;
    }
    await this.recordTurnAsync(message, reply, prepared.queryEmbedding, prepared.snippets);
  }

  async agentAsync(instruction: string): Promise<string> {
    // Pilot: no tool-execution loop yet — agentic tool calling is a later slice.
    // Falls back to a plain reply so the surface is complete.
    return this.sendAsync(instruction);
  }

  getContext(): CompanionContext {
    return this._context;
  }

  async refreshContextAsync(): Promise<void> {
    const hits = await this.recall.recallAsync("", null, this.recallTopK());
    this._context = this.buildContext(hits.map((h) => h.item.text));
  }

  async signalFeedbackAsync(_positive: boolean, _note?: string): Promise<void> {
    // Pilot: feedback is accepted but not yet routed to a feedback store / affect
    // update. Wired in a later slice.
  }

  // ── internals ──────────────────────────────────────────────────────────────

  private async prepareAsync(
    message: string,
  ): Promise<{ messages: ChatMessage[]; queryEmbedding: number[] | null; snippets: string[] }> {
    // Recall runs BEFORE the current turn is persisted, so it draws on prior
    // memory, never echoes the message back.
    const queryEmbedding = this.opts.embedder ? await this.opts.embedder(message) : null;
    const hits = await this.recall.recallAsync(message, queryEmbedding, this.recallTopK());
    const snippets = hits.map((h) => h.item.text);

    const messages: ChatMessage[] = [{ role: "system", content: this.buildSystemPrompt(snippets) }];
    for (const turn of this._history) messages.push({ role: turn.role, content: turn.content });
    messages.push({ role: "user", content: message });

    return { messages, queryEmbedding, snippets };
  }

  private async recordTurnAsync(
    userText: string,
    reply: string,
    queryEmbedding: number[] | null,
    snippets: string[],
  ): Promise<void> {
    const episodeId = randomUUID();
    const entry: EpisodicMemoryEntry = {
      id: episodeId,
      recordedAtUtc: new Date(),
      userText,
      assistantText: reply,
      appContext: this.opts.appContext,
      embedding: queryEmbedding ?? undefined,
    };
    await this.episodic.addAsync(entry);

    // Off the hot path: fill the graph + form attributed beliefs for next time.
    this.opts.encoder?.enqueue(userText, reply, episodeId);

    const now = new Date();
    this._history.push({ role: "user", content: userText, timestamp: now });
    this._history.push({ role: "assistant", content: reply, timestamp: now });
    this._context = this.buildContext(snippets);
  }

  private buildSystemPrompt(snippets: string[]): string {
    const parts: string[] = [];
    if (this.opts.personaHints) parts.push(this.opts.personaHints.trim());
    if (this.opts.affectSummary) parts.push(this.opts.affectSummary.trim());

    const facts = this.userFacts();
    if (facts.length > 0) {
      parts.push("[What you know about the user]\n" + facts.map((f) => "- " + f).join("\n"));
    }
    if (snippets.length > 0) {
      parts.push("[Relevant memories]\n" + snippets.map((s) => "- " + s).join("\n"));
    }
    return parts.join("\n\n");
  }

  private userFacts(): string[] {
    if (!this.opts.beliefs) return [];
    return this.opts.beliefs.selfFacts().map((f) => f.object);
  }

  private buildContext(snippets: string[]): CompanionContext {
    return {
      identityId: this.identityId,
      displayName: this.opts.displayName ?? "",
      preferredLanguage: this.opts.preferredLanguage ?? null,
      interface: this.interface,
      personaHints: this.opts.personaHints ?? "",
      affectSummary: this.opts.affectSummary ?? "",
      recentMemorySnippets: snippets,
      activeGoals: this.opts.activeGoals ?? [],
      contextBuiltAt: new Date(),
    };
  }

  private recallTopK(): number {
    return this.opts.recallTopK ?? 5;
  }
}

File: rust/src/companion/session.rs
//! session.rs
//!
//! The conscious loop: a concrete [`ICompanionSession`] that recalls from fused
//! memory, persists each turn, and encodes it into the graph off the hot path.
//! Ported from CircleAI.Companion (CompanionSession) — the C# reference — and
//! mirrors the TypeScript pilot (companion/session.ts) and the Go port
//! (companion_session.go) 1:1.
//!
//! On every turn it (1) recalls the most relevant memories + the user's own facts
//! and injects them into the system prompt, (2) calls the generator, (3) persists
//! the exchange to episodic memory, and (4) hands it to the background encoder so
//! the knowledge graph fills for future associative recall.

use std::sync::Arc;

use chrono::Utc;
use uuid::Uuid;

use crate::brain::BrainError;
use crate::companion::belief::SelfBeliefStore;
use crate::companion::memory_encoder::CompanionMemoryEncoder;
use crate::companion::types::{
    CompanionContext, CompanionTurn, ICompanionSession, InterfaceKind,
};
use crate::inference::IChatGenerator;
use crate::memory::episodic::InMemoryEpisodicStore;
use crate::memory::graph::MemoryHit;
use crate::memory::recall::IRecall;
use crate::memory::stores::EpisodicMemoryEntry;
use crate::models::ChatMessage;

/// Computes an embedding for the given text. Returns `None` when no embedding is
/// available (→ episodic recency recall). Optional.
pub type EmbedderFn =
    dyn Fn(&str) -> Result<Option<Vec<f32>>, BrainError> + Send + Sync;

/// Construction-time configuration for a [`CompanionSession`].
#[derive(Default)]
pub struct CompanionSessionOptions {
    pub session_id: String,
    pub identity_id: String,
    pub interface: InterfaceKind,
    pub display_name: String,
    pub preferred_language: Option<String>,
    /// A static persona hint block prepended to the system prompt.
    pub persona_hints: String,
    /// A static affect hint block prepended to the system prompt.
    pub affect_summary: String,
    pub active_goals: Vec<String>,
    /// How many memories to recall per turn. Default 5 (when 0).
    pub recall_top_k: usize,
    /// An optional app context stamped onto persisted episodes.
    pub app_context: Option<String>,
    /// The background graph/belief encoder. When `None`, turns are not encoded.
    pub encoder: Option<Arc<CompanionMemoryEncoder>>,
    /// Holds the user's own facts, surfaced into the system prompt.
    pub beliefs: Option<Arc<SelfBeliefStore>>,
    /// An optional embedder for associative episodic recall; `None` → recency.
    pub embedder: Option<Arc<EmbedderFn>>,
}

impl Default for InterfaceKind {
    fn default() -> Self {
        InterfaceKind::Headless
    }
}

/// A companion session that thinks with fused memory and remembers what it
/// learns. Generic over the concrete chat generator `G` (reuses the existing
/// [`IChatGenerator`] trait); tests inject a fake generator that captures the
/// messages it is handed.
pub struct CompanionSession<G: IChatGenerator> {
    generator: G,
    episodic: Arc<InMemoryEpisodicStore>,
    recall: Arc<dyn IRecall>,
    opts: CompanionSessionOptions,

    history: Vec<CompanionTurn>,
    context: CompanionContext,
}

struct PreparedTurn {
    messages: Vec<ChatMessage>,
    query_embedding: Option<Vec<f32>>,
    snippets: Vec<String>,
}

impl<G: IChatGenerator> CompanionSession<G> {
    /// Creates a session. `generator`, `episodic` and `recall` are required.
    pub fn new(
        generator: G,
        episodic: Arc<InMemoryEpisodicStore>,
        recall: Arc<dyn IRecall>,
        opts: CompanionSessionOptions,
    ) -> Result<Self, BrainError> {
        let context = build_context(&opts, &[]);
        Ok(Self {
            generator,
            episodic,
            recall,
            opts,
            history: Vec::new(),
            context,
        })
    }

    fn recall_top_k(&self) -> usize {
        if self.opts.recall_top_k == 0 {
            5
        } else {
            self.opts.recall_top_k
        }
    }

    /// Borrows the underlying chat generator. Useful for tests that inject a
    /// capturing generator and want to inspect the messages it was handed.
    pub fn generator(&self) -> &G {
        &self.generator
    }

    /// Recalls before the current turn is persisted, so recall draws on prior
    /// memory and never echoes the message back.
    fn prepare(&self, message: &str) -> Result<PreparedTurn, BrainError> {
        let query_embedding = match &self.opts.embedder {
            Some(embed) => embed(message)?,
            None => None,
        };

        let hits = self.recall.recall(
            message,
            query_embedding.as_deref(),
            self.recall_top_k(),
        )?;
        let snippets = snippets_from_hits(&hits);

        let mut messages: Vec<ChatMessage> =
            vec![ChatMessage::system(self.build_system_prompt(&snippets))];
        for turn in &self.history {
            messages.push(ChatMessage::new(turn.role.clone(), turn.content.clone()));
        }
        messages.push(ChatMessage::user(message));

        Ok(PreparedTurn {
            messages,
            query_embedding,
            snippets,
        })
    }

    fn record_turn(
        &mut self,
        user_text: &str,
        reply: &str,
        query_embedding: Option<Vec<f32>>,
        snippets: Vec<String>,
    ) -> Result<(), BrainError> {
        let episode_id = Uuid::new_v4();
        let entry = EpisodicMemoryEntry {
            id: episode_id,
            recorded_at_utc: Utc::now(),
            user_text: user_text.to_string(),
            assistant_text: reply.to_string(),
            app_context: self.opts.app_context.clone(),
            embedding: query_embedding,
            tags: None,
        };
        self.episodic.add_shared(entry)?;

        // Off the hot path: fill the graph + form attributed beliefs for next time.
        if let Some(encoder) = &self.opts.encoder {
            encoder.enqueue(user_text, reply, &episode_id.to_string());
        }

        let now = Utc::now();
        self.history.push(CompanionTurn {
            role: "user".to_string(),
            content: user_text.to_string(),
            timestamp: now,
        });
        self.history.push(CompanionTurn {
            role: "assistant".to_string(),
            content: reply.to_string(),
            timestamp: now,
        });
        self.context = build_context(&self.opts, &snippets);
        Ok(())
    }

    fn build_system_prompt(&self, snippets: &[String]) -> String {
        let mut parts: Vec<String> = Vec::new();
        if !self.opts.persona_hints.trim().is_empty() {
            parts.push(self.opts.persona_hints.trim().to_string());
        }
        if !self.opts.affect_summary.trim().is_empty() {
            parts.push(self.opts.affect_summary.trim().to_string());
        }

        let facts = self.user_facts();
        if !facts.is_empty() {
            let mut b = String::from("[What you know about the user]");
            for f in &facts {
                b.push_str("\n- ");
                b.push_str(f);
            }
            parts.push(b);
        }
        if !snippets.is_empty() {
            let mut b = String::from("[Relevant memories]");
            for snip in snippets {
                b.push_str("\n- ");
                b.push_str(snip);
            }
            parts.push(b);
        }
        parts.join("\n\n")
    }

    fn user_facts(&self) -> Vec<String> {
        match &self.opts.beliefs {
            None => Vec::new(),
            Some(beliefs) => beliefs
                .self_facts()
                .into_iter()
                .map(|f| f.object)
                .collect(),
        }
    }
}

impl<G: IChatGenerator> ICompanionSession for CompanionSession<G>
where
    G::Error: 'static,
{
    type Error = BrainError;

    fn session_id(&self) -> &str {
        &self.opts.session_id
    }

    fn identity_id(&self) -> &str {
        &self.opts.identity_id
    }

    fn interface(&self) -> InterfaceKind {
        self.opts.interface
    }

    fn send(&mut self, message: &str) -> Result<String, Self::Error> {
        let prepared = self.prepare(message)?;
        let reply = self
            .generator
            .generate(&prepared.messages, None)
            .map_err(|e| BrainError::new(e.to_string()))?;
        self.record_turn(
            message,
            &reply,
            prepared.query_embedding,
            prepared.snippets,
        )?;
        Ok(reply)
    }

    fn stream(
        &mut self,
        message: &str,
    ) -> Result<Box<dyn Iterator<Item = Result<String, Self::Error>>>, Self::Error> {
        let prepared = self.prepare(message)?;

        // Drive the generator's stream to completion, accumulating the reply so
        // the full text is persisted once the stream ends (mirrors the reference:
        // stream chunks out, persist the accumulated reply on completion).
        let stream = self
            .generator
            .stream(&prepared.messages, None)
            .map_err(|e| BrainError::new(e.to_string()))?;

        let mut chunks: Vec<String> = Vec::new();
        let mut accumulated = String::new();
        for item in stream {
            let chunk = item.map_err(|e| BrainError::new(e.to_string()))?;
            accumulated.push_str(&chunk);
            chunks.push(chunk);
        }

        self.record_turn(
            message,
            &accumulated,
            prepared.query_embedding,
            prepared.snippets,
        )?;

        Ok(Box::new(chunks.into_iter().map(Ok)))
    }

    fn agent(&mut self, instruction: &str) -> Result<String, Self::Error> {
        // Pilot: no tool-execution loop yet — agentic tool calling is a later
        // slice. Falls back to a plain reply so the surface is complete.
        self.send(instruction)
    }

    fn get_context(&self) -> &CompanionContext {
        &self.context
    }

    fn refresh_context(&mut self) -> Result<(), Self::Error> {
        let hits = self.recall.recall("", None, self.recall_top_k())?;
        let snippets = snippets_from_hits(&hits);
        self.context = build_context(&self.opts, &snippets);
        Ok(())
    }

    fn history(&self) -> &[CompanionTurn] {
        &self.history
    }

    fn signal_feedback(&mut self, _positive: bool, _note: Option<&str>) -> Result<(), Self::Error> {
        // Pilot: accepted but not yet routed to a feedback store / affect update.
        Ok(())
    }
}

fn build_context(opts: &CompanionSessionOptions, snippets: &[String]) -> CompanionContext {
    CompanionContext {
        identity_id: opts.identity_id.clone(),
        display_name: opts.display_name.clone(),
        preferred_language: opts.preferred_language.clone(),
        interface: opts.interface,
        persona_hints: opts.persona_hints.clone(),
        affect_summary: opts.affect_summary.clone(),
        recent_memory_snippets: snippets.to_vec(),
        active_goals: opts.active_goals.clone(),
        context_built_at: Utc::now(),
    }
}

fn snippets_from_hits(hits: &[MemoryHit]) -> Vec<String> {
    hits.iter().map(|h| h.item.text.clone()).collect()
}
